File: backend/dispatch/src/main/java/com/commonsengine/dispatch/api/DispatchController.java
package com.commonsengine.dispatch.api;

import com.commonsengine.dispatch.domain.Model.DispatchTask;
import com.commonsengine.dispatch.domain.Model.RouteSuggestion;
import com.commonsengine.dispatch.domain.Model.TaskStatus;
import com.commonsengine.dispatch.domain.Model.WorkerPreferences;
import com.commonsengine.dispatch.service.DispatchService;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Instant;
import java.util.List;
import java.util.Set;

/**
 * 调度 REST API（#66）
 *
 * 端点：
 * - POST   /api/v1/dispatch/tasks                 创建派单任务
 * - GET    /api/v1/dispatch/tasks/pending         查询待分配任务
 * - POST   /api/v1/dispatch/tasks/{id}/assign     分配任务
 * - POST   /api/v1/dispatch/preferences           保存劳动者偏好
 * - GET    /api/v1/dispatch/preferences/{workerId} 查询偏好
 * - POST   /api/v1/dispatch/optimize-route         路径优化
 */
@RestController
@RequestMapping("/api/v1/dispatch")
public class DispatchController {

    private final DispatchService service;

    public DispatchController(DispatchService service) {
        this.service = service;
    }

    @PostMapping("/tasks")
    public ResponseEntity<DispatchTaskResponse> createTask(
            @RequestBody DispatchTaskRequest request) {
        DispatchTask t = service.createTask(request.serviceType(), request.pickup().lat(),
                request.pickup().lng());
        return ResponseEntity.status(HttpStatus.CREATED).body(DispatchTaskResponse.from(t));
    }

    @GetMapping("/tasks/pending")
    public List<DispatchTaskResponse> pendingTasks() {
        return service.findPendingTasks().stream()
                .map(DispatchTaskResponse::from)
                .toList();
    }

    @PostMapping("/tasks/{id}/assign")
    public DispatchTaskResponse assign(@PathVariable String id,
                                        @RequestBody AssignRequest request) {
        DispatchTask t = service.assignTask(id, request.workerId());
        return DispatchTaskResponse.from(t);
    }

    @PostMapping("/preferences")
    public WorkerPreferencesResponse savePreferences(
            @RequestBody WorkerPreferencesRequest request) {
        WorkerPreferences p = new WorkerPreferences(
                request.workerId(),
                request.serviceTypes(),
                request.workingHours(),
                request.maxServiceRadiusMeters()
        );
        WorkerPreferences saved = service.savePreferences(p);
        return WorkerPreferencesResponse.from(saved);
    }

    @GetMapping("/preferences/{workerId}")
    public WorkerPreferencesResponse getPreferences(@PathVariable String workerId) {
        return WorkerPreferencesResponse.from(service.findPreferences(workerId));
    }

    @PostMapping("/optimize-route")
    public RouteSuggestionResponse optimizeRoute(@RequestBody OptimizeRouteRequest request) {
        List<double[]> points = request.visits().stream()
                .map(v -> new double[]{v.lat(), v.lng()})
                .toList();
        RouteSuggestion suggestion = service.optimizeRoute(
                request.origin().lat(),
                request.origin().lng(),
                points,
                request.workerId()
        );
        return RouteSuggestionResponse.from(suggestion);
    }

    // ── DTO records ────────────────────────────────────────

    public record GeoPointDto(double lat, double lng) {
    }

    public record TimeSlotDto(Instant start, Instant end) {
    }

    public record DispatchTaskRequest(
            @NotBlank String serviceType,
            @NotNull GeoPointDto pickup
    ) {
    }

    public record DispatchTaskResponse(
            String id,
            String serviceType,
            GeoPointDto pickup,
            String assignedWorkerId,
            TaskStatus status,
            Instant createdAt
    ) {
        static DispatchTaskResponse from(DispatchTask t) {
            return new DispatchTaskResponse(
                    t.id(),
                    t.serviceType(),
                    new GeoPointDto(t.pickupLat(), t.pickupLng()),
                    t.assignedWorkerId(),
                    t.status(),
                    t.createdAt()
            );
        }
    }

    public record WorkerPreferencesRequest(
            @NotBlank String workerId,
            Set<String> serviceTypes,
            com.commonsengine.dispatch.domain.Model.TimeSlot workingHours,
            double maxServiceRadiusMeters
    ) {
    }

    public record WorkerPreferencesResponse(
            String workerId,
            Set<String> serviceTypes,
            com.commonsengine.dispatch.domain.Model.TimeSlot workingHours,
            double maxServiceRadiusMeters
    ) {
        static WorkerPreferencesResponse from(WorkerPreferences p) {
            return new WorkerPreferencesResponse(
                    p.workerId(), p.serviceTypes(), p.workingHours(), p.maxServiceRadiusMeters()
            );
        }
    }

    public record OptimizeRouteRequest(
            @NotBlank String workerId,
            @NotNull GeoPointDto origin,
            List<GeoPointDto> visits
    ) {
    }

    public record RouteSuggestionResponse(
            String workerId,
            List<VisitDto> visits,
            double totalDistanceMeters,
            long estimatedDurationMinutes,
            String strategy
    ) {
        static RouteSuggestionResponse from(RouteSuggestion s) {
            List<VisitDto> visits = s.visits().stream()
                    .map(v -> new VisitDto(v.sequence(), v.lat(), v.lng(), v.taskId(), v.label()))
                    .toList();
            return new RouteSuggestionResponse(s.workerId(), visits,
                    s.totalDistanceMeters(), s.estimatedDurationMinutes(), s.strategy());
        }
    }

    public record VisitDto(int sequence, double lat, double lng, String taskId, String label) {
    }

    public record AssignRequest(@NotBlank String workerId) {
    }
}

File: backend/dispatch/src/main/kotlin/com/commonsengine/dispatch/api/DispatchController.kt
package com.commonsengine.dispatch.api

import com.commonsengine.dispatch.domain.DispatchTask
import com.commonsengine.dispatch.domain.TimeSlot
import com.commonsengine.dispatch.domain.WorkerPreferences
import com.commonsengine.dispatch.service.DispatchService
import com.commonsengine.platform.domain.ServiceType
import com.commonsengine.platform.geo.GeoPoint
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

/**
 * 调度引擎 REST API
 *
 * 设计原则（架构文档 3.5 条）：
 * 1. 路径优化服务于劳动者效率，而非平台抽成最大化
 * 2. 劳动者可设定工作偏好和边界，引擎尊重这些设定
 */
@RestController
@RequestMapping("/api/v1/dispatch")
open class DispatchController(
    private val service: DispatchService,
) {

    /** 分配调度任务 */
    @PostMapping("/tasks")
    fun assignTask(@RequestBody body: DispatchTaskRequest): DispatchTaskResponse {
        val task = DispatchTask(
            id = body.id ?: UUID.randomUUID().toString(),
            workerId = body.workerId,
            serviceType = ServiceType.valueOf(body.serviceType),
            pickups = body.pickups.map { GeoPoint(it.lat, it.lng) },
            dropoffs = body.dropoffs.map { GeoPoint(it.lat, it.lng) },
            estimatedDistanceMeters = body.estimatedDistanceMeters ?: 0.0,
            estimatedDurationMinutes = body.estimatedDurationMinutes ?: 0,
            deadline = null,
        )
        val saved = service.assignTask(task)
        return saved.toResponse()
    }

    /** 查询调度任务 */
    @GetMapping("/tasks/{taskId}")
    fun findTask(@PathVariable taskId: String): DispatchTaskResponse? =
        service.findTask(taskId)?.toResponse()

    /** 查询某劳动者的所有调度任务 */
    @GetMapping("/workers/{workerId}/tasks")
    fun findTasksByWorker(@PathVariable workerId: String): List<DispatchTaskResponse> =
        service.findTasksByWorker(workerId).map { it.toResponse() }

    /** 保存或更新劳动者工作偏好 */
    @PostMapping("/workers/{workerId}/preferences")
    fun savePreferences(
        @PathVariable workerId: String,
        @RequestBody body: WorkerPreferencesRequest,
    ): Map<String, String> {
        val prefs = WorkerPreferences(
            workerId = workerId,
            preferredServiceTypes = body.preferredServiceTypes
                .mapNotNull { runCatching { ServiceType.valueOf(it) }.getOrNull() }.toSet(),
            preferredRegions = body.preferredRegions.toSet(),
            excludedRegions = body.excludedRegions.toSet(),
            preferredTimeSlots = body.preferredTimeSlots.map { it.toTimeSlot() }.toSet(),
            excludedTimeSlots = body.excludedTimeSlots.map { it.toTimeSlot() }.toSet(),
            maxConcurrentOrders = body.maxConcurrentOrders ?: 3,
            maxDailyHours = body.maxDailyHours ?: 12.0,
        )
        service.savePreferences(prefs)
        return mapOf("status" to "ok", "workerId" to workerId)
    }

    /** 查询劳动者工作偏好 */
    @GetMapping("/workers/{workerId}/preferences")
    fun findPreferences(@PathVariable workerId: String): WorkerPreferencesResponse? {
        val prefs = service.findPreferences(workerId) ?: return null
        return WorkerPreferencesResponse(
            workerId = prefs.workerId,
            preferredServiceTypes = prefs.preferredServiceTypes.map { it.name },
            preferredRegions = prefs.preferredRegions.toList(),
            excludedRegions = prefs.excludedRegions.toList(),
            preferredTimeSlots = prefs.preferredTimeSlots.map { TimeSlotDto(it.dayOfWeek, it.startHour, it.endHour) },
            excludedTimeSlots = prefs.excludedTimeSlots.map { TimeSlotDto(it.dayOfWeek, it.startHour, it.endHour) },
            maxConcurrentOrders = prefs.maxConcurrentOrders,
            maxDailyHours = prefs.maxDailyHours,
        )
    }

    /** 优化路径 */
    @PostMapping("/optimize-route")
    fun optimizeRoute(@RequestBody body: OptimizeRouteRequest): RouteSuggestionResponse {
        val task = DispatchTask(
            id = "optimize-${UUID.randomUUID()}",
            workerId = body.workerId,
            serviceType = ServiceType.valueOf(body.serviceType),
            pickups = body.pickups.map { GeoPoint(it.lat, it.lng) },
            dropoffs = body.dropoffs.map { GeoPoint(it.lat, it.lng) },
        )
        val currentLocation = GeoPoint(body.currentLat, body.currentLng)
        val route = service.optimizeRoute(currentLocation, task)
        return RouteSuggestionResponse(
            workerId = route.workerId,
            orderedWaypoints = route.orderedWaypoints.map { GeoPointDto(it.lat, it.lng) },
            totalDistanceMeters = route.totalDistanceMeters,
            estimatedTotalMinutes = route.estimatedTotalMinutes,
            reason = route.reason,
        )
    }
}

// ── DTO ──────────────────────────────────────────────────────────

data class GeoPointDto(val lat: Double, val lng: Double)

data class TimeSlotDto(val dayOfWeek: Int, val startHour: Int, val endHour: Int) {
    fun toTimeSlot() = TimeSlot(dayOfWeek, startHour, endHour)
}

data class DispatchTaskRequest(
    val id: String? = null,
    val workerId: String,
    val serviceType: String,
    val pickups: List<GeoPointDto>,
    val dropoffs: List<GeoPointDto>,
    val estimatedDistanceMeters: Double? = null,
    val estimatedDurationMinutes: Int? = null,
)

data class DispatchTaskResponse(
    val id: String,
    val workerId: String,
    val serviceType: String,
    val pickupCount: Int,
    val dropoffCount: Int,
    val estimatedDistanceMeters: Double,
    val estimatedDurationMinutes: Int,
    val assignedAt: String,
)

data class WorkerPreferencesRequest(
    val preferredServiceTypes: List<String> = emptyList(),
    val preferredRegions: List<String> = emptyList(),
    val excludedRegions: List<String> = emptyList(),
    val preferredTimeSlots: List<TimeSlotDto> = emptyList(),
    val excludedTimeSlots: List<TimeSlotDto> = emptyList(),
    val maxConcurrentOrders: Int? = null,
    val maxDailyHours: Double? = null,
)

data class WorkerPreferencesResponse(
    val workerId: String,
    val preferredServiceTypes: List<String>,
    val preferredRegions: List<String>,
    val excludedRegions: List<String>,
    val preferredTimeSlots: List<TimeSlotDto>,
    val excludedTimeSlots: List<TimeSlotDto>,
    val maxConcurrentOrders: Int,
    val maxDailyHours: Double,
)

data class OptimizeRouteRequest(
    val workerId: String,
    val serviceType: String,
    val currentLat: Double,
    val currentLng: Double,
    val pickups: List<GeoPointDto>,
    val dropoffs: List<GeoPointDto>,
)

data class RouteSuggestionResponse(
    val workerId: String,
    val orderedWaypoints: List<GeoPointDto>,
    val totalDistanceMeters: Double,
    val estimatedTotalMinutes: Int,
    val reason: String,
)

private fun DispatchTask.toResponse() = DispatchTaskResponse(
    id = id,
    workerId = workerId,
    serviceType = serviceType.name,
    pickupCount = pickups.size,
    dropoffCount = dropoffs.size,
    estimatedDistanceMeters = estimatedDistanceMeters,
    estimatedDurationMinutes = estimatedDurationMinutes,
    assignedAt = assignedAt.toString(),
)
